packages/api/endpoints/rules.js

```diff
@@ -123,29 +137,33 @@ async function put({ params: { name }, body }, res) {
   }
 
   try {
-    const oldRule = await model.get({ name });
-    const dbClient = await getKnexClient();
-    const rulePgModel = new RulePgModel();
+    const oldRule = await ruleModel.get({ name });
 
     apiRule.updatedAt = Date.now();
     apiRule.createdAt = oldRule.createdAt;
-    // If rule type is onetime no change is allowed unless it is a rerun
 
+    // If rule type is onetime no change is allowed unless it is a rerun
     if (apiRule.action === 'rerun') {
       return models.Rule.invoke(oldRule).then(() => res.send(oldRule));
     }
 
     const fieldsToDelete = Object.keys(oldRule).filter(
       (key) => !(key in apiRule) && key !== 'createdAt'
     );
-    const postgresRule = await translateApiRuleToPostgresRule(apiRule, dbClient);
-
-    await dbClient.transaction(async (trx) => {
-      await rulePgModel.upsert(trx, postgresRule);
-      newRule = await model.update(oldRule, apiRule, fieldsToDelete);
-    });
+    const postgresRule = await translateApiRuleToPostgresRule(apiRule, knex);
+
+    try {
+      await knex.transaction(async (trx) => {
+        await rulePgModel.upsert(trx, postgresRule);
+        newRule = await ruleModel.update(oldRule, apiRule, fieldsToDelete);
+        await indexRule(esClient, newRule, process.env.ES_INDEX);
+      });
+    } catch (innerError) {
+      // Revert Dynamo record update if any write fails
+      await ruleModel.create(oldRule);
```

Comment:
Same as above. If `await rulePgModel.upsert(trx, postgresRule);` fails I think this would try to recreate an existing rule. This would apply to all `put` endpoints.

Reply:
So it's true that if `rulePgModel.upsert` fails, then it is going to run `ruleModel.create(oldRule)` to restore the original Dynamo record. Truthfully, this is unnecessary if it's the PG write that failed, because it never updated the Dynamo record. However, it shouldn't be harmful to issue a request to restore the original record even if it was never updated.

It's important to note that our Dynamo model `create` methods are actually doing a PUT under the hood, so they will either create a record or update an existing one.

Thoughts?

Reply:
That makes sense. What if `indexRule` fails though? In that case will we throw and overwrite the updated rule with the old one? If yes, would we also want to do the same for PG? Or if `indexRule` fails will it not throw 🤔?

Reply:
> In that case will we throw and overwrite the updated rule with the old one?

Yes, exactly

> If yes, would we also want to do the same for PG

We don't need to manually overwrite for PG, because the PG write only gets committed to the database if everything in the transaction succeeds. 

> Or if indexRule fails will it not throw

I would expect it to throw if there is any error

There is a test for the `if Elasticsearch write fails` here:

https://github.com/nasa/cumulus/blob/e586d77c0cd38756ccca92c149e03e7228c80d4c/packages/api/tests/endpoints/test-rules.js#L881

And if Dynamo write fails:

https://github.com/nasa/cumulus/blob/e586d77c0cd38756ccca92c149e03e7228c80d4c/packages/api/tests/endpoints/test-rules.js#L753

And if PG write fails:

https://github.com/nasa/cumulus/blob/e586d77c0cd38756ccca92c149e03e7228c80d4c/packages/api/tests/endpoints/test-rules.js#L819

Reply:
Awesome. Thanks for taking the time to step through it. Seems totally sane.